exercises/src/main/scala/fpinscala/state/State.scala

```diff
@@ -30,32 +30,127 @@ object RNG {
       (f(a), rng2)
     }
 
-  def nonNegativeInt(rng: RNG): (Int, RNG) = ???
+  def map_via_flatMap[A,B](s: Rand[A])(f: A => B): Rand[B] =
+    flatMap(s)(a => unit(f(a)))
 
-  def double(rng: RNG): (Double, RNG) = ???
+  def nonNegativeInt(rng: RNG): (Int, RNG) = {
+    val (i, r) = rng.nextInt
+    (if (i < 0) -(i + 1) else i, r)
+  }
 
-  def intDouble(rng: RNG): ((Int,Double), RNG) = ???
+  def double(rng: RNG): (Double, RNG) = {
+    val (i, r) = nonNegativeInt(rng)
+    ((i.toDouble)/(Int.MaxValue.toDouble+1), r)
+  }
 
-  def doubleInt(rng: RNG): ((Double,Int), RNG) = ???
+  def double_via_map: Rand[Double] =
+    map(nonNegativeInt)(i => (i.toDouble)/(Int.MaxValue.toDouble+1))
 
-  def double3(rng: RNG): ((Double,Double,Double), RNG) = ???
+  def intDouble(rng: RNG): ((Int,Double), RNG) = {
+    val (i, rng2) = rng.nextInt
+    val (d, rng3) = double(rng2)
+    ((i,d), rng3)
+  }
 
-  def ints(count: Int)(rng: RNG): (List[Int], RNG) = ???
+  def doubleInt(rng: RNG): ((Double,Int), RNG) = {
+    val ((i,d), rng2) = intDouble(rng)
+    ((d,i), rng2)
+  }
 
-  def map2[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = ???
+  def double3(rng: RNG): ((Double,Double,Double), RNG) = {
+    val (d1, rng2) = double(rng)
+    val (d2, rng3) = double(rng2)
+    val (d3, rng4) = double(rng3)
+    ((d1,d2,d3), rng4)
+  }
 
-  def sequence[A](fs: List[Rand[A]]): Rand[List[A]] = ???
+  def ints(count: Int)(rng: RNG): (List[Int], RNG) = {
+    @annotation.tailrec
+    def go(c: Int, acc: List[Int])(rng: RNG): (List[Int], RNG) = {
+      if (c < 1) (acc, rng)
+      else {
+        val (i, r) = rng.nextInt
+        go(c-1, i::acc)(r)
+      }
+    }
+    go(count, List())(rng)
+  }
+
+  def ints_via_sequence(count: Int)(rng: RNG): (List[Int], RNG) =
+    sequence(List.fill(count)(nonNegativeInt(_)))(rng)
+
+  def map2[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = 
+    rng => {
+      val (a, rngA) = ra(rng)
+      val (b, rngB) = rb(rngA)
+      (f(a, b), rngB)
+    }
 
-  def flatMap[A,B](f: Rand[A])(g: A => Rand[B]): Rand[B] = ???
+  def map2_via_flatMap[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = 
+    //flatMap(ra)(a => f(a,rb(r)))
+    flatMap(ra)(a => map(rb)(b => f(a,b)))
+
+  def sequence[A](fs: List[Rand[A]]): Rand[List[A]] = 
+    fs.foldRight(unit(List[A]()))((h,t) => map2(h, t)(_ :: _))
+
+  def nonNegativeLessThan(n: Int): Rand[Int] =
+    flatMap(nonNegativeInt)(i => if (i-i%n+n-1 >= 0) unit(i%n) else nonNegativeLessThan(n))
+
+  def flatMap[A,B](f: Rand[A])(g: A => Rand[B]): Rand[B] = 
+    rng => {
+      val (a, r2) = f(rng)
+      val (b, r3) = g(a)(r2)
+      (b, r3)
+    }
 }
 
 case class State[S,+A](run: S => (A, S)) {
-  def map[B](f: A => B): State[S, B] =
-    sys.error("todo")
+  def map[B](f: A => B): State[S, B] = 
+    // State { (s: S) =>
+    //   val (a, s2) = run(s)
+    //   (f(a), s2)
+    // }
+    flatMap(a => State.unit(f(a)))
```

Comment:
Can also be written as `flatMap(f andThen State.unit)`. This definition holds true for any type with `unit`, `flatMap`, and `map` operations.